src/saf/saf_base.nit

```diff
@@ -0,0 +1,236 @@
+	fun pretty_print is abstract
+end
+
+# An analysis go forward from the entry point to the exit point.
+#
+# With a forward analysis, output properties are determined by the inputs.
+class ForwardAnalysis
+	super StaticAnalysis
+
+	redef fun visit(n) do n.accept_forward_analysis(self)
+end
+
+# FlowSet are used to represent data at the entry and exit of a statement.
+interface FlowSet
+	super Cloneable
+
+	# Merge `self` with another flow set.
+	fun flow_union(o: SELF): SELF is abstract
+end
+
+# A FlowSet based on a HashMap.
+class FlowHashMap[K, V]
+	super FlowSet
+	super HashMap[K, V]
+
+	# Init `self` with the content of `map`.
+	init from(map: Map[K, V]) do
+		init
+		for k, v in map do self[k] = v
+	end
+
+	# Not a deep copy.
+	redef fun clone do return new FlowHashMap[K, V].from(self)
+end
+
+# A FlowSet based on a HashSet.
+class FlowHashSet[E]
+	super FlowSet
+	super HashSet[E]
+
+	redef fun clone do return new FlowHashSet[E].from(self)
+
+	# Remove all items found in `other` from `self`.
+	fun remove_from(other: Collection[E]) do
+		for e in other do remove(e)
+	end
+
+	redef fun flow_union(o) do return new FlowHashSet[E].from(union(o))
+end
+
+redef class ANode
+
+	# Apply the forward analysis `v` to `self`.
+	fun accept_forward_analysis(v: ForwardAnalysis) do
+		v.current_inset = v.current_outset.clone
+		v.current_outset = v.current_inset.clone
+		v.insets[self] = v.current_inset
+		visit_all(v)
+		v.outsets[self] = v.current_outset
+	end
+end
+
+redef class AIfExpr
+
+	# Merge flow on if .. else constructs.
+	redef fun accept_forward_analysis(v) do
+		v.enter_visit(n_expr)
+		var inset = v.current_inset
+		var outset = v.current_outset
+
+		if n_then != null then v.enter_visit(n_then)
+		var then_outset = v.current_outset
+
+		v.current_inset = inset
+		v.current_outset = outset
+
+		if n_else != null then
+			v.enter_visit(n_else)
+			outset = v.merge(then_outset, v.current_outset)
+		else
+			outset = v.merge(then_outset, v.current_inset)
+		end
+		v.current_inset = inset
+		v.current_outset = outset
+	end
+end
+
+# Represent all kind of loop nodes.
+#
+# Used to factorize implementations across whiles, fors and loops.
+interface ALoopHelper
+	# Lookup fix point for this loop.
+	fun loop_fix_point(v: StaticAnalysis, node: ANode) do
+		var inset = v.current_inset.clone
+		var last: nullable FlowSet = null
+		while v.current_outset != last do
+			v.enter_visit(node)
+			v.current_inset = v.merge(inset, v.current_outset)
+			v.current_outset = v.current_inset.clone
+			last = v.current_outset.clone
+		end
+		v.current_inset = inset
+		v.current_outset = v.merge(inset, v.current_outset)
+	end
+
+	# Factorize loop forward analysis.
+	fun accept_loop_forward_analysis(v: StaticAnalysis) do
+		var n_block = loop_block
+		if not n_block == null then loop_fix_point(v, n_block)
+	end
+
+	# The block contained by this loop.
+	fun loop_block: nullable ANode is abstract
+end
+
+redef class ADoExpr
+	super ALoopHelper
```

Comment:
Yes but it has a block inside with a sequence of instructions, I think this is similar to a loop (from a factorization point of view)